### PR TITLE
Add comics api client with tests.

### PR DESCRIPTION
**Pull request by flipper83**
- Add comic api client.
- It's missing change the code of the demo, but it's out of scope of this PR.

### CODE REVIEW

File: MarvelApiClient/src/main/java/com/karumi/marvelapiclient/ComicApiClient.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ */
+
+package com.karumi.marvelapiclient;
+
+import com.karumi.marvelapiclient.model.CharacterDto;
+import com.karumi.marvelapiclient.model.CharactersDto;
+import com.karumi.marvelapiclient.model.CharactersQuery;
```
Comment (pedrovgs) [inline MarvelApiClient/src/main/java/com/karumi/marvelapiclient/ComicApiClient.java]: You have non used imports, please reformat all the code related to this PR and review the imports.

**Comment by pedrovgs** [comment]
We should create two packages, one for `CharactersApiClient` and another one for `ComicsApiClient` and move also all the DTO related to these packages.

File: MarvelApiClient/src/main/java/com/karumi/marvelapiclient/model/ComicsQuery.java [hunk 2]
```diff
@@ -0,0 +1,574 @@
...
+      returnValues.put(QUERY_ORDER_BY, orderBy);
+    }
+
+    if (limit > 0) {
+      returnValues.put(QUERY_LIMIT, limit);
+    }
+
+    if (offset > 0) {
+      returnValues.put(QUERY_OFFSET, offset);
+    }
```
Comment (pedrovgs) [inline MarvelApiClient/src/main/java/com/karumi/marvelapiclient/model/ComicsQuery.java:224]: What's the problem adding null values to the map? Retrofit should not send null values. In the case of the limit or the offset you can declare it as `Integer` and avoid all these non null checks.

File: MarvelApiClient/src/test/java/com/karumi/marvelapiclient/CharacterApiClientTest.java [hunk 3]
```diff
@@ -86,7 +87,7 @@
         .withLimit(ANY_LIMIT)
         .withOffset(ANY_OFFSET)
         .withModifiedSince(getAnyDate())
-        .withOrderBy(CharactersQuery.OrderBy.NAME, false)
+        .withOrderBy(OrderBy.NAME, false)
```
Comment (pedrovgs) [inline MarvelApiClient/src/test/java/com/karumi/marvelapiclient/CharacterApiClientTest.java:90]: You have a non used method in `CharactersQuery` you should connect with the new one. This is the method:

```
 public Builder withOrderBy(OrderBy orderBy) {
      this.orderBy = orderBy;
      this.orderByAscendant = true;
      return this;
    }
```

File: MarvelApiClient/src/main/java/com/karumi/marvelapiclient/ComicApiClient.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
+/*
+ *   Copyright (C) 2015 Karumi.
```
Comment (Serchinastico) [inline MarvelApiClient/src/main/java/com/karumi/marvelapiclient/ComicApiClient.java:2]: Is it me or is this copyright missing an empty line? :P

File: MarvelApiClient/src/test/java/com/karumi/marvelapiclient/ComicApiClientTest.java [hunk 5]
```diff
@@ -0,0 +1,351 @@
...
+  private static final String ANY_MODIFIED_SINCE = "2015-01-09T22:10:45-0800";
+  private static final String ORDER_NAME_DESCEDANT_VALUE = "-name";
+  private static final String INVALID_COMIC_ID = "";
+  private static final String ANY_NOT_FOUND_ID = "1234";
+  private static final String ANY_COMIC_ID = "123456";
+  public static final String EVENTS_REQUEST = "1,2";
+  public static final String COLLABORATORS_REQUEST = "1,2";
+  private static final String LAST_WEEK_REQUEST = "last%20week";
+
+  @Test public void shouldRequestComicsByOffsetAndLimitWithAsParams() throws Exception {
```
Comment (pedrovgs) [inline MarvelApiClient/src/test/java/com/karumi/marvelapiclient/ComicApiClientTest.java:82]: should**Send**Request.... Please, review this and the rest of methods using the word `Request` before the `should` word.

File: MarvelApiClient/src/test/java/com/karumi/marvelapiclient/ComicApiClientTest.java [hunk 6]
```diff
@@ -0,0 +1,351 @@
...
+  @Test public void shouldSendGetComicRequestToTheCorrectEndpoint() throws Exception {
+    ComicApiClient comicApiClient = givenComicApiClient();
+    enqueueMockResponse("getComic.json");
+
+    comicApiClient.getComic(ANY_COMIC_ID);
+
+    assertRequestSentToContains("comics/" + ANY_COMIC_ID);
+  }
+
+  @Test(expected = MarvelApiException.class)
```
Comment (pedrovgs) [inline MarvelApiClient/src/test/java/com/karumi/marvelapiclient/ComicApiClientTest.java:189]: Don't we have a `ComicNotFoundException`?

**Comment by pedrovgs** [comment]
About your comment: `It's missing change the code of the demo, but it's out of scope of this PR.` Why don't you remove the current demo code based on a CLI application and replace it with just a main java method requesting hardcoded data? This is easy to maintain and shows how to use the library easily.

File: MarvelApiClient/src/main/java/com/karumi/marvelapiclient/model/ComicsQuery.java [hunk 7]
```diff
@@ -0,0 +1,574 @@
...
+      if (limit <= 0) {
+        throw new IllegalArgumentException("limit must be bigger than zero");
+      }
+
+      if (limit > MAX_SIZE) {
+        throw new IllegalArgumentException("limit must be smaller than 100");
+      }
+    }
+
+    private void checkNull(Object object) {
```
Comment (Serchinastico) [inline MarvelApiClient/src/main/java/com/karumi/marvelapiclient/model/ComicsQuery.java]: `checkNotNull`?

File: MarvelApiClient/src/main/java/com/karumi/marvelapiclient/ComicApiClient.java (hunk 1 above)
Comment (flipper83) [inline MarvelApiClient/src/main/java/com/karumi/marvelapiclient/ComicApiClient.java]: yep

File: MarvelApiClient/src/main/java/com/karumi/marvelapiclient/model/ComicsQuery.java (hunk 2 above)
Comment (flipper83) [inline MarvelApiClient/src/main/java/com/karumi/marvelapiclient/model/ComicsQuery.java:224]: I have problems with the marvel api and nulls, some nulls appear on the url, for this reason I made this null check.

File: MarvelApiClient/src/main/java/com/karumi/marvelapiclient/model/ComicsQuery.java (hunk 7 above)
Comment (flipper83) [inline MarvelApiClient/src/main/java/com/karumi/marvelapiclient/model/ComicsQuery.java]: yep

File: MarvelApiClient/src/main/java/com/karumi/marvelapiclient/model/ComicsQuery.java (hunk 2 above)
Comment (pedrovgs) [inline MarvelApiClient/src/main/java/com/karumi/marvelapiclient/model/ComicsQuery.java:224]: So Retrofit is not ignoring null values for the url encoding?

File: MarvelApiClient/src/test/java/com/karumi/marvelapiclient/CharacterApiClientTest.java (hunk 3 above)
Comment (flipper83) [inline MarvelApiClient/src/test/java/com/karumi/marvelapiclient/CharacterApiClientTest.java:90]: yep.

File: MarvelApiClient/src/test/java/com/karumi/marvelapiclient/ComicApiClientTest.java (hunk 6 above)
Comment (flipper83) [inline MarvelApiClient/src/test/java/com/karumi/marvelapiclient/ComicApiClientTest.java:189]: nop

Comment (pedrovgs) [inline MarvelApiClient/src/test/java/com/karumi/marvelapiclient/ComicApiClientTest.java:189]: Should we create one to tell the library user that comic does not exist?

File: MarvelApiClient/src/main/java/com/karumi/marvelapiclient/model/ComicsQuery.java (hunk 2 above)
Comment (flipper83) [inline MarvelApiClient/src/main/java/com/karumi/marvelapiclient/model/ComicsQuery.java:224]: Yep, I had problems with that.

Comment (pedrovgs) [inline MarvelApiClient/src/main/java/com/karumi/marvelapiclient/model/ComicsQuery.java:224]: ok

**Pull request merged**